rust/src/quic/quic.rs

```diff
@@ -112,44 +122,111 @@ impl QuicState {
+            match QuicHeader::from_bytes(buf, DEFAULT_DCID_LEN) {
+                Ok((rest, header)) => {
+                    // unprotect
+                    if self.keys.is_none() && header.ty == QuicType::Initial {
+                        if let Some(version) = quic_rustls_version(u32::from(header.version)) {
+                            let keys = rustls::quic::Keys::initial(version, &header.dcid, false);
+                            self.keys = Some(keys)
+                        }
+                    }
+                    if header.length > rest.len() {
+                        //TODO event whenever an error condition is met
+                        return false;
+                    }
+                    let (next_buf, mut framebuf) = (&rest[header.length..], &rest[..header.length]);
+                    let mut rest2;
+                    if let Some(keys) = &self.keys {
+                        let hkey = if to_server {
+                            &keys.remote.header
+                        } else {
+                            &keys.local.header
+                        };
+                        if framebuf.len() < 4 + hkey.sample_len() {
+                            return false;
+                        }
+                        let hlen = buf.len() - rest.len();
+                        let mut h2 = Vec::with_capacity(hlen + header.length);
+                        h2.extend_from_slice(&buf[..hlen + header.length]);
+                        let mut h20 = h2[0];
+                        let mut pktnum_buf = Vec::with_capacity(4);
+                        pktnum_buf.extend_from_slice(&h2[hlen..hlen + 4]);
+                        let r1 = hkey.decrypt_in_place(
+                            &h2[hlen + 4..hlen + 4 + hkey.sample_len()],
+                            &mut h20,
+                            &mut pktnum_buf,
+                        );
+                        if !r1.is_ok() {
+                            return false;
+                        }
+                        // mutate one at a time
+                        h2[0] = h20;
+                        let _ = &h2[hlen..hlen + 4].copy_from_slice(&pktnum_buf);
+                        let pkt_num = quic_pkt_num(&h2[hlen..hlen + 1 + ((h20 & 3) as usize)]);
+                        rest2 = Vec::with_capacity(framebuf.len() - (1 + ((h20 & 3) as usize)));
+                        if framebuf.len() < 1 + ((h20 & 3) as usize) {
+                            return false;
+                        }
+                        rest2.extend_from_slice(&framebuf[1 + ((h20 & 3) as usize)..]);
+                        let pkey = if to_server {
+                            &keys.remote.packet
+                        } else {
+                            &keys.local.packet
+                        };
+                        let r = pkey.decrypt_in_place(pkt_num, &h2[..hlen + 4], &mut rest2);
+                        if !r.is_ok() {
+                            return false;
```

Comment:
As for SSH, should we stop parsing because everything is encrypted at some point ?

Reply:
I think not, I think it should be more like tls/tcp, where we can continue to inspect the tls itself w/o inspecting the encrypted data in the detect engine